src/python/pants/backend/scala/dependency_inference/ScalaParser.scala

```diff
@@ -223,8 +229,18 @@ class SourceAnalysisTraverser extends Traverser {
     case Defn.Object(mods, nameNode, templ) => {
       visitMods(mods)
       val name = extractName(nameNode)
-      recordProvidedName(name, sawObject = true)
-      visitTemplate(templ, name)
+
+      // TODO: should object already be recursive?
+      // an object is recursive if extends another type because we cannot figure out the provided types
+      // in the parents, we just mark the object as recursive (which is indicated by non-empty inits)
+      val recursive = !templ.inits.isEmpty
+      recordProvidedName(name, sawObject = true, recursive = recursive)
+      
+      // If the object is recursive, no need to provide the symbols inside
```

Comment:
What if `object B extends A` put is in a separate file or adds additional functions, etc? Wouldn't those have to be marked as provided?

Reply:
Because object B is now recursive, the dependency inference code will add a dependency on the file of object B regardless if it can find a match. The same way third-party dependencies work.

code is here:
https://github.com/pantsbuild/pants/blob/main/src/python/pants/jvm/dependency_inference/artifact_mapper.py#L165

Reply:
That makes sense. Cool!